include/fiction/algorithms/physical_design/graph_oriented_layout_design.hpp

```diff
@@ -0,0 +1,1646 @@
+    DEST
+};
+/**
+ * Struct to hold information necessary for gate placement during layout generation for one vertex.
+ *
+ * @tparam ObstrLyt The type of the layout.
+ */
+template <typename ObstrLyt>
+struct placement_info
+{
+    /**
+     * The index of the current node being placed.
+     */
+    uint64_t current_node;
+    /**
+     * The index of the current primary output.
+     */
+    uint64_t current_po;
+    /**
+     * Mapping of nodes to their positions in the layout.
+     */
+    node_dict_type<ObstrLyt, tec_nt> node2pos;
+    /**
+     * Mapping of primary input nodes to layout nodes.
+     */
+    mockturtle::node_map<mockturtle::node<ObstrLyt>, tec_nt> pi2node;
+};
+/**
+ * Implementation of the graph-oriented layout design algorithm.
+ * This class handles the initialization and execution of the algorithm.
+ *
+ * @tparam Lyt Cartesian gate-level layout type.
+ * @tparam Ntk Network type.
+ */
+template <typename Lyt, typename Ntk>
+class graph_oriented_layout_design_impl
+{
+  public:
+    /**
+     * Constructor for the graph-oriented layout design algorithm.
+     *
+     * @param src The source network to be placed.
+     * @param p The parameters for the graph-enhanced layout search algorithm.
+     * @param st The statistics object to record execution details.
+     */
+    graph_oriented_layout_design_impl(const Ntk& src, const graph_oriented_layout_design_params& p,
+                                      graph_oriented_layout_design_stats& st) :
+            ntk{convert_network<tec_nt>(src)},
+            ps{p},
+            pst{st},
+            timeout{ps.timeout},
+            num_vertex_expansions{ps.num_vertex_expansions},
+            start{std::chrono::high_resolution_clock::now()},
+            num_search_space_graphs{ps.high_effort_mode ? 12u : 2u},
+            ssg_vec(num_search_space_graphs)
+    {
+        ntk.substitute_po_signals();
+    }
+    /**
+     * Executes the graph-oriented layout design algorithm and returns the best found layout.
+     *
+     * @return The best layout found by the algorithm.
+     */
+    std::optional<Lyt> run() noexcept
+    {
+        // measure run time
+        mockturtle::stopwatch stop{pst.time_total};
+
+        // initialize layout to keep track of current best solution
+        Lyt best_lyt{{}, twoddwave_clocking<Lyt>()};
+
+        // initialize search space graphs
+        initialize();
+
+        // check if a timeout was set
+        bool timeout_set = (timeout != std::numeric_limits<uint64_t>::max());
+
+        // if no timeout was set and high-effort mode is disabled, set timeout to 10s
+        if (!timeout_set && !ps.high_effort_mode)
+        {
+            timeout = 10000u;
+        }
+        bool timeout_limit_reached = false;
+
+#if (PROGRESS_BARS)
+        // initialize a progress bar
+        mockturtle::progress_bar bar{ssg_vec[0].network.size(), "[i] arranging layout: |{0}|"};
+        uint64_t                 max_placed_nodes_tracker = 0u;
```

Comment:
There are some problems with the progress bar as of right now. It immediately displays a completed bar.